source/Camera/InfoUtilOther/C++Halcon/HalconConnectUsingGenICam/HalconConnectUsingGenICam.cpp

```diff
@@ -0,0 +1,102 @@
+#include <Zivid/Zivid.h>
+#include <halconcpp/HalconCpp.h>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+#include <chrono>
+#include <iostream>
+
+
+void savePointCloud(const HalconCpp::HObjectModel3D &model, const std::string filename)
+{
+	const HalconCpp::HString filetype{ "ply" };
+	const HalconCpp::HString filenameCstring{ filename.c_str() };
+	const HalconCpp::HString genParamName{ "invert_normals" };
+	const HalconCpp::HString genParamValue{ "false" };
+	model.WriteObjectModel3d(filetype, filenameCstring, genParamName, genParamValue);
+}
+
+
+HalconCpp::HObjectModel3D setColorInObjectModel3D(HalconCpp::HImage RGB, HalconCpp::HObjectModel3D pcHalcon) {
+
+	HalconCpp::HImage R = RGB.AccessChannel(1);
+	HalconCpp::HImage G = RGB.AccessChannel(2);
+	HalconCpp::HImage B = RGB.AccessChannel(3);
+
+	HalconCpp::HRegion domain = R.GetDomain();
+	HalconCpp::HTuple rows, cols;
+	domain.GetRegionPoints(&rows, &cols);
+
+	HalconCpp::HTuple Rtup = R.GetGrayval(rows, cols);
+	HalconCpp::HTuple Gtup = G.GetGrayval(rows, cols);
+	HalconCpp::HTuple Btup = B.GetGrayval(rows, cols);
```

Comment:
Mest of these can be `const`. Also, you can get rid of R, G, B variables and use RGB directly.

```
const HalconCpp::HRegion domain = RGB.AccessChannel(1).GetDomain();
HalconCpp::HTuple rows, cols;
domain.GetRegionPoints(&rows, &cols);

const HalconCpp::HTuple Rtup = RGB.AccessChannel(1).GetGrayval(rows, cols);
const HalconCpp::HTuple Gtup = RGB.AccessChannel(2).GetGrayval(rows, cols);
const HalconCpp::HTuple Btup = RGB.AccessChannel(3).GetGrayval(rows, cols);
```
Also I am not sure if `domain.GetRegionPoints(&rows, &cols);` is needed at all? Does it do anything? I don't understand how a 'get' function returns void. Is it modifying rows and cols?

Are you sure indexing starts from 1 and not from 0? Seems weid for C++ but it is Halcon so anything is possible.

Reply:
How do you think I can do this differently?

GetRegionPoints returns the region data in the form of a coordinate list. I'm not entirely sure how Halcon structures an Image, but I need rows and cols to access the gray values. And yeah, indexing starts from 1. Halcon is not C++ "friendly", so there are several things that do not make sense to do in a C++ way of thinking.

Reply:
Based on your explanation there is no other way. That is fine. I just didn't' understand how their function works.

Reply:
no one understands how they work, hehe